### PR TITLE
Allow merge of configs referencing the same domain

### LINKED ISSUE #206: hot reload with a same domain is not supported.
l am using `RUNTIME_WATCH_ROOT` using hot reload the configuration. Is it possible to update the same domain name ?

level=error msg="error loading new configuration from runtime: config-1.yaml: config file cannot have empty domain"
time="2021-01-15T11:30:13Z" level=error msg="error loading new configuration from runtime: config.yaml: duplicate domain 'vhost-ratelimit' in config file"
time="2021-01-15T11:30:13Z" level=error msg="error loading new configuration from runtime: config.yaml: duplicate domain 'vhost-ratelimit' in config file"
time="2021-01-15T11:30:13Z" level=error msg="error loading new configuration from runtime: config.yaml: duplicate domain 'vhost-ratelimit' in config file"
time="2021-01-15T11:31:22Z" level=error msg="error loading new configuration from runtime: config-1.yaml: config file cannot have empty domain"

### PR DESCRIPTION
**Pull request by m-rcl**
Add MERGE_DOMAIN_CONFIG variable defaulting to false and mapping to the former behavior.
Setting to true will allow multiple files containing non duplicate keys and referencing the same domain to be merged.

* Fixes #206
* Continues #209

### CODE REVIEW

**Comment by m-rcl** [comment]
@mattklein123 @ysawa0 could one of you please have a look? Thanks in advance!

File: src/config_check_cmd/main.go [hunk 1]
```diff
@@ -15,21 +15,23 @@ import (
...
 	defer func() {
 		err := recover()
 		if err != nil {
 			fmt.Printf("error loading rate limit configs: %s\n", err.(error).Error())
 			os.Exit(1)
 		}
 	}()
-	statsManager := stats.NewStatManager(gostats.NewStore(gostats.NewNullSink(), false), settings.NewSettings())
-	config.NewRateLimitConfigImpl(allConfigs, statsManager)
+	statsManager := stats.NewStatManager(gostats.NewStore(gostats.NewNullSink(), mergeDomainConfigs), settings.NewSettings())
```
Comment (mattklein123) [inline src/config_check_cmd/main.go:26]: I think this is a typo?

File: src/service/ratelimit.go [hunk 2]
```diff
@@ -75,12 +75,15 @@ func (this *service) reloadConfig(statsManager stats.Manager) {
 		files = append(files, config.RateLimitConfigToLoad{key, snapshot.Get(key)})
 	}
 
-	newConfig := this.configLoader.Load(files, statsManager)
+	this.configLock.Lock()
+	defer this.configLock.Unlock()
```
Comment (mattklein123) [inline src/service/ratelimit.go:79]: For performance reasons I think you want to keep the locking where it is. I think you can just move the settings construction outside the lock? I'm not sure why it's in there. Seems like that should be static anyway?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small comments, thank you.

File: src/service/ratelimit.go (hunk 2 above)
Comment (m-rcl) [reply src/service/ratelimit.go]: Agree. Having a second look at it, moving out the constructor should not have any impact.

File: src/config_check_cmd/main.go (hunk 1 above)
Comment (mattklein123) [inline src/config_check_cmd/main.go:26]:
```suggestion
	statsManager := stats.NewStatManager(gostats.NewStore(gostats.NewNullSink(), false), settings.NewSettings())
```

Not sure why this line changed?

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**